src/hotspot/share/prims/jvm.cpp

```diff
@@ -4044,3 +4044,22 @@ JVM_ENTRY(void, JVM_VirtualThreadUnmountEnd(JNIEnv* env, jobject vthread, jboole
   fatal("Should only be called with JVMTI enabled");
 #endif
 JVM_END
+
+/*
+ * Return the current class's class file version.  The low order 16 bits of the
+ * the returned jint contains the class's major version.  The high order 16 bits
+ * contains the class's minor version.
+ */
+JVM_ENTRY(jint, JVM_GetClassFileVersion(JNIEnv* env, jclass current))
+  oop mirror = JNIHandles::resolve_non_null(current);
+  if (java_lang_Class::is_primitive(mirror)) {
+    // return latest major version and minor version of 0.
+    return JVM_CLASSFILE_MAJOR_VERSION;
+  }
+  assert(!java_lang_Class::as_Klass(mirror)->is_array_klass(), "unexpected array class");
```

Comment:
Can this throw IllegalArgumentException instead.
Asserts only report problems when built with debug (Right?)

Reply:
Or, Can the VM do this traversal as/more efficiently than doing at the java level?

Reply:
It is usual for the Java code to do such checks and throw exceptions, so that the VM assumes it is correct and only asserts incase something has been missed on the Java side.

Reply:
Though in this case the Java code has defined behaviour for array types so it is correct for the VM to assume this is not an array type and to assert if it is.

Reply:
My companion question was whether the native code can for efficiently map from array class to component type than Java.

Reply:
I don't know whether Java or the JVM is faster but I would rather keep this code in Java so that I don't have to change existing method JVM_GetClassAccessFlags().  (It's used by Class.getClassAccessFlagsRaw().)

Reply:
Makes sense, re-using the existing method.